File: Java/PrisonerSimulator/src/Game/State/DetailedGameState.java
package Game.State;

import Game.Game;

import java.util.ArrayList;

public class DetailedGameState extends ProcessGameState {
    private int round;

    private void PrintCompetitionStepInfo() {
        var player = currentCompetitors.get(currentPlayerIndex);
        var rival = currentCompetitors.get(currentRivalIndex);

        System.out.println("Step complete");

        var playerLastDecision = player.getLastDecision(rival.getId()).toString();
        int playerDelta = player.getLastDelta(rival.getId());
        System.out.printf(" %s: %sm Delta: %d (Score = %d)\n", player.getName(), playerLastDecision, playerDelta, player.getScore());

        var rivalLastDecision = rival.getLastDecision(player.getId()).toString();
        int rivalDelta = rival.getLastDelta(player.getId());
        System.out.printf(" %s: %sm Delta: %d (Score = %d)\n", rival.getName(), rivalLastDecision, rivalDelta, rival.getScore());
    }

    public DetailedGameState(Game game) {
        super(game);
        currentCompetitors = new ArrayList<>(game.getPlayers());
        round = 0;
    }

    @Override
    public boolean doStage() {
        RunCompetitionStep();
        PrintCompetitionStepInfo();

        currentPlayerIndex += 1;
        currentRivalIndex = (currentRivalIndex + 1) % currentCompetitors.size();

        boolean isRoundComplete = currentPlayerIndex == currentCompetitors.size();
        if (isRoundComplete) {
            currentPlayerIndex = 0;
            currentRivalIndex = 1;

            System.out.printf("Round %d finished\n\n", round);
            ++round;

            boolean isGameFinished = round == game.getGameRules().getRoundsCount();
            if (isGameFinished) {
                game.switchGameState(new EndGameState(game));
            }
        }

        return false;
    }
}

File: Kotlin/PrisonerSimulator/src/game/state/DetailedGameState.kt
package game.state

import game.Game

class DetailedGameState(game: Game) : ProcessGameState(game) {
    override val currentCompetitors = game.players
    private var round = 0

    private fun printCompetitionStepInfo() {
        val (player, rival) = currentCompetitors[currentPlayerIndex] to currentCompetitors[currentRivalIndex]
        val (playerDecision, rivalDecision) = player.getLastDecision(rival.id) to rival.getLastDecision(player.id)

        println("Step complete")
        println("${player.name}: $playerDecision Delta: ${player.getLastDelta(rival.id)} (Score = ${player.score})")
        println("${rival.name}: $rivalDecision Delta: ${rival.getLastDelta(player.id)} (Score = ${rival.score})")
    }

    override fun doStage(): Boolean {
        runCompetitionStep()
        printCompetitionStepInfo()

        ++currentPlayerIndex
        currentRivalIndex = (currentRivalIndex + 1) % currentCompetitors.size

        val isRoundComplete = currentPlayerIndex == currentCompetitors.size
        if (isRoundComplete) {
            currentPlayerIndex = 0
            currentRivalIndex = 1
            println("Round $round finished\n\n")
            ++round

            val isGameFinished = round == game.gameRules.roundsCount
            if (isGameFinished) {
                game.gameState = EndGameState(game)
            }
        }

        return false
    }
}
